src/main/java/org/ice4j/stack/MessageProcessingTask.java

```diff
@@ -0,0 +1,213 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ice4j.stack;
+
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.*;
+import java.util.function.*;
+import java.util.logging.*;
+
+import org.ice4j.*;
+import org.ice4j.message.*;
+
+/**
+ * The class is used to parse and dispatch incoming messages by being
+ * executed by concurrent {@link java.util.concurrent.ForkJoinPool}.
+ * To reduce memory allocation this <tt>ForkJoinTask</tt> implementation
+ * designed to be suitable for usage with pooling, the instance of this type is
+ * mutable such that <tt>RawMessage</tt> can be updated and instance can be
+ * reused and scheduled with new <tt>RawMessage</tt>
+ *
+ * @author Emil Ivov
+ */
+class MessageProcessingTask
+    extends RecursiveAction
+{
+    /**
+     * Our class logger.
+     */
+    private static final Logger logger
+        = Logger.getLogger(MessageProcessingTask.class.getName());
+
+    /**
+     * Indicates that <tt>MessageProcessingTask</tt> is cancelled and should not
+     * process <tt>RawMessage</tt> anymore.
+     */
+    private final AtomicBoolean cancelled = new AtomicBoolean(false);
+
+    /**
+     * The <tt>NetAccessManager</tt> which has created this instance and which
+     * is its owner.
+     */
+    private final NetAccessManager netAccessManager;
+
+    /**
+     * The listener that will be collecting error notifications.
+     */
+    private final ErrorHandler errorHandler;
+
+    /**
+     * The listener that will be retrieving <tt>MessageEvent</tt>s
+     */
+    private final MessageEventHandler messageEventHandler;
+
+    /**
+     * Raw message which is being processed
+     */
+    private RawMessage rawMessage;
+
+    /**
+     * Callback which is invoked when this <tt>MessageProcessingTask</tt>
+     * processed it's {@link #rawMessage}
+     */
+    private Consumer<MessageProcessingTask> rawMessageProcessedHandler;
+
+    /**
+     * Creates a Message processor.
+     *
+     * @param netAccessManager the <tt>NetAccessManager</tt> which is creating
+     * the new instance, is going to be its owner, specifies the
+     * <tt>MessageEventHandler</tt> and represents the <tt>ErrorHandler</tt> to
+     * handle exceptions in the new instance
+     * @throws IllegalArgumentException if any of the mentioned properties of
+     * <tt>netAccessManager</tt> are <tt>null</tt>
+     */
+    MessageProcessingTask(
+        NetAccessManager netAccessManager)
+        throws IllegalArgumentException
+    {
+        if (netAccessManager == null)
```

Comment:
It was here before, one reason I see it was added is when `NullPointerException` generated it is immediately clear which variable is `null`, otherwise the only way to find out would be to open code at specific version which is currently deployed and find out throwing line by line number from stack trace, something like `NetAccessManager.java:95`.